src/main/java/software/amazon/kinesis/connectors/flink/internals/publisher/polling/PollingRecordPublisher.java

```diff
@@ -108,13 +110,27 @@ public RecordPublisherRunResult run(final RecordBatchConsumer consumer, int maxN
 		GetRecordsResult result = getRecords(nextShardItr, maxNumberOfRecords);
 
 		RecordBatch recordBatch = new RecordBatch(result.getRecords(), subscribedShard, result.getMillisBehindLatest());
-		SequenceNumber latestSeequenceNumber = consumer.accept(recordBatch);
 
-		nextStartingPosition = StartingPosition.continueFromSequenceNumber(latestSeequenceNumber);
+		SequenceNumber latestSequenceNumber = consumer.accept(recordBatch);
+
+		nextStartingPosition = getNextStartingPosition(latestSequenceNumber);
 		nextShardItr = result.getNextShardIterator();
 		return nextShardItr == null ? COMPLETE : INCOMPLETE;
 	}
 
+	private StartingPosition getNextStartingPosition(final SequenceNumber latestSequenceNumber) {
+		// When consuming from a timestamp sentinel/AT_TIMESTAMP ShardIteratorType.
+		// If the first RecordBatch is empty, then the latestSequenceNumber would be the timestamp sentinel.
+		// This is because we have not yet received any real sequence numbers on this shard.
+		// In this condition we should retry from the previous starting position (AT_TIMESTAMP).
+		if (SENTINEL_AT_TIMESTAMP_SEQUENCE_NUM.get().equals(latestSequenceNumber)) {
```

Comment:
Sadly, `SentinelSequenceNumber#sentinel` is not declared as final, so I think it's safer to explicitly precompute `SENTINEL_AT_TIMESTAMP_SEQUENCE_NUM.get()` by pulling it out of the `getNextStartingPosition` (this will be on the hot path) instead of relying on the JIT compiler.

Reply:
Fixed